src/libstd/sync/mutex.rs

```diff
@@ -463,14 +439,47 @@ impl<T: ?Sized + fmt::Display> fmt::Display for MutexGuard<'_, T> {
     }
 }
 
-pub fn guard_lock<'a, T: ?Sized>(guard: &MutexGuard<'a, T>) -> &'a sys::Mutex {
-    &guard.__lock.inner
+pub fn guard_lock<'a, 'b, T: ?Sized>(guard: &'b mut MutexGuard<'a, T>)
+    -> &'b mut parking_lot::MutexGuard<'a, T>
+{
+    &mut guard.__inner
 }
 
 pub fn guard_poison<'a, T: ?Sized>(guard: &MutexGuard<'a, T>) -> &'a poison::Flag {
     &guard.__lock.poison
 }
 
+
+/// A parking_lot low level mutex without data in it.
+/// This type is not exposed from the standard library, only used inside it.
+pub struct RawMutex(parking_lot::RawMutex);
```

Comment:
What is the point of not using parking_lot's `RawMutex` directly instead?

Reply:
Because it does not have a corresponding `RawMutexGuard` that perform automatic unlocking on drops, so it's more low level. This small helper makes using it more ergonomic throughout the standard library. Most places where this new `RawMutex` is now used previously used an equivalent `crate::sys_common::mutex::Mutex;`. If we don't have this helper type in place we will need to introduce a number of `LOCK.unlock()` calls here and there instead of just holding a `_guard` that will be dropped.

Reply:
IIRC the only reason a `RawMutex` type was used instead of `Mutex<()>` was because the latter boxed the real mutex type. However now that the mutex isn't boxed any more, there is no reason not to use `Mutex<()>` directly.

Reply:
Because most places that use this `RawMutex` needs the constructor to be `const`. Which `parking_lot::Mutex` isn't. And can't be for now. So I had to introduce this `RawMutex` to get mutex with const initialization and guards.

Reply:
Later, when types that **has generics** can be `const` **in the standard library**, we can remove the `RawMutex` in favor for just `parking_lot::Mutex<()>` indeed.

I think there were some possible workaround to this where we made the `parking_lot` module a staged API. But I'm not familiar with that and did not want to dig so deep into it back when I was working on that.

Reply:
Generic functions can be const. They just cannot have bounds, they must work for *all* types.

Reply:
Ah, that was probably the limitation, yes! I forgot the details, it was a while back.

Reply:
Does that mean we could use `Mutex<()>` instead of introducing a `RawMutex` with guards?

Reply:
In any case please add a comment saying that unlike `parking_lot::RawMutex`, this has guards.

Reply:
No we can't. Because `parking_lot::Mutex` has trait bounds. It's actually just a type def for the mutex in `lock_api`:
```rust
pub type Mutex<T> = lock_api::Mutex<RawMutex, T>;
```
Which has trait bounds: https://github.com/Amanieu/parking_lot/blob/master/lock_api/src/mutex.rs#L99

But yes, I'll add docs on `RawMutex` declaring why it exists and that it can be removed when the ordinary mutex can have a const constructor.